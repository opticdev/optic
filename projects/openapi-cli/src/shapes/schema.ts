import { OpenAPIV3 } from '../specs/index';
import { ShapePatch } from './patches';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { Operation } from '../patches';
import JsonPatch from 'fast-json-patch';

export type SchemaObject = OpenAPIV3.SchemaObject;

export class Schema {
  static baseFromValue(value: any): SchemaObject {
    const rootSchema = initialSchema(value);

    let schema = rootSchema;

    return schema;
  }

  static clone(value: SchemaObject): SchemaObject {
    return JsonPatch.deepClone(value);
  }

  static merge(
    currentSchema: SchemaObject,
    newSchema: SchemaObject
  ): SchemaObject {
    const merged = { ...currentSchema, ...newSchema };

    let allowedKeys: string[] = [...allowedMetaDataForAll, 'type'];
    if (newSchema.type === 'object') allowedKeys = allowedKeysForObject;
    if (newSchema.type === 'array') allowedKeys = allowedKeysForArray;
    if (newSchema.type === 'string') allowedKeys = allowedKeysForString;
    if (newSchema.type === 'number' || newSchema.type === 'integer')
      allowedKeys = allowedKeysForInteger;
    if (newSchema.oneOf) allowedKeys = allowedKeysForOneOf;

    for (let key in merged) {
      if (!allowedKeys.includes(key) || isExtension(key)) {
        delete merged[key];
      }
    }

    return merged;
  }

  static *mergeOperations(
    currentSchema: SchemaObject,
    newSchema: SchemaObject
  ): IterableIterator<Operation> {
    const merged = { ...currentSchema, ...newSchema };
    const currrentKeys = new Set(Object.keys(currentSchema));

    let allowedKeys: string[] = [...allowedMetaDataForAll, 'type'];
    if (newSchema.type === 'object') allowedKeys = allowedKeysForObject;
    if (newSchema.type === 'array') allowedKeys = allowedKeysForArray;
    if (newSchema.type === 'string') allowedKeys = allowedKeysForString;
    if (newSchema.type === 'number' || newSchema.type === 'integer')
      allowedKeys = allowedKeysForInteger;
    if (newSchema.oneOf) allowedKeys = allowedKeysForOneOf;

    for (let [key, value] of Object.entries(merged)) {
      let path = jsonPointerHelpers.append('', key);

      if (!allowedKeys.includes(key) || isExtension(key)) {
        yield {
          op: 'remove',
          path,
        };
      } else if (!currrentKeys.has(key)) {
        yield {
          op: 'add',
          path,
          value,
        };
      } else if (value !== currentSchema[value]) {
        yield {
          op: 'replace',
          path,
          value,
        };
      }
    }
  }

  static applyShapePatch(
    schema: SchemaObject,
    patch: ShapePatch
  ): SchemaObject {
    const result = JsonPatch.applyPatch(
      schema,
      [...ShapePatch.operations(patch)],
      undefined,
      false // don't mutate the original schema
    );

    return result.newDocument;
  }
}

function initialSchema(rootInput: any): OpenAPIV3.SchemaObject {
  if (rootInput === null) {
    // @ts-ignore, this is ok now.
    return { type: 'null' };
  } else if (Array.isArray(rootInput)) {
    return {
      type: 'array',
      items: rootInput.length ? initialSchema(rootInput[0]) : {},
    };
  } else if (typeof rootInput === 'object') {
    return { type: 'object' };
  } else if (typeof rootInput === 'string') {
    return { type: 'string' };
  } else if (typeof rootInput === 'number') {
    return { type: 'number' };
  } else if (typeof rootInput === 'boolean') {
    return { type: 'boolean' };
  } else {
    throw new Error('Could not learn JSON Schema');
  }
}

export const allowedMetaDataForAll: string[] = [
  'title',
  'description',
  'example',
  'examples',
  'default',
  'deprecated',
  'externalDocs',
];

export const allowedKeysForOneOf: string[] = [
  ...allowedMetaDataForAll,
  'oneOf',
  'allOf',
  'not',
  'discriminator',
];

export const allowedKeysForObject: string[] = [
  ...allowedMetaDataForAll,
  'additionalProperties',
  'type',
  'maxProperties',
  'minProperties',
  'required',
  'properties',
];
export const allowedKeysForArray: string[] = [
  ...allowedMetaDataForAll,
  'type',
  'maxItems',
  'minItems',
  'uniqueItems',
];

export const allowedKeysForString: string[] = [
  ...allowedMetaDataForAll,
  'type',
  'format',
  'pattern',
  'maxLength',
  'minLength',
];

export const allowedKeysForInteger: string[] = [
  ...allowedMetaDataForAll,
  'type',
  'minimum',
  'maximum',
  'multipleOf',
];

export const isExtension = (key: string) => key.startsWith('x-');
