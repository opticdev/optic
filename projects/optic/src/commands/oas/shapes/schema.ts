import { OpenAPIV3 } from '../specs/index';
import { ShapePatch } from './patches';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { PatchOperation } from '../patches';
import JsonPatch from 'fast-json-patch';
import { OperationPatch } from '../operations';
import { SupportedOpenAPIVersions } from '@useoptic/openapi-io';

export type SchemaObject = OpenAPIV3.SchemaObject;

export class Schema {
  static baseFromValue(
    value: any,
    openAPIVersion: SupportedOpenAPIVersions
  ): SchemaObject {
    const rootSchema = initialSchema(value, openAPIVersion);

    let schema = rootSchema;

    return schema;
  }

  static clone(value: SchemaObject): SchemaObject {
    return JsonPatch.deepClone(value);
  }

  static equals(a: SchemaObject, b: SchemaObject): boolean {
    return !!Schema.mergeOperations(a, b).next().done;
  }

  static merge(
    currentSchema: SchemaObject,
    newSchema: SchemaObject
  ): SchemaObject {
    const merged = { ...currentSchema, ...newSchema };

    let allowedKeys: string[] = [...allowedMetaDataForAll, 'type'];
    if (newSchema.type === 'object') allowedKeys = allowedKeysForObject;
    if (newSchema.type === 'array') allowedKeys = allowedKeysForArray;
    if (newSchema.type === 'string') allowedKeys = allowedKeysForString;
    if (newSchema.type === 'number' || newSchema.type === 'integer')
      allowedKeys = allowedKeysForInteger;
    if (newSchema.oneOf) allowedKeys = allowedKeysForOneOf;

    for (let key in merged) {
      if (!allowedKeys.includes(key) || isExtension(key)) {
        delete merged[key];
      }
    }

    return merged;
  }

  static *mergeOperations(
    currentSchema: SchemaObject | null,
    newSchema: SchemaObject
  ): IterableIterator<PatchOperation> {
    if (currentSchema === null) {
      yield {
        op: 'add',
        path: '',
        value: {},
      };
      currentSchema = {};
    }

    const merged = { ...currentSchema, ...newSchema };
    const currentKeys = new Set(Object.keys(currentSchema));

    let allowedKeys: string[] = [...allowedMetaDataForAll, 'type'];
    if (newSchema.type === 'object') allowedKeys = allowedKeysForObject;
    if (newSchema.type === 'array') allowedKeys = allowedKeysForArray;
    if (newSchema.type === 'string') allowedKeys = allowedKeysForString;
    if (newSchema.type === 'number' || newSchema.type === 'integer')
      allowedKeys = allowedKeysForInteger;
    if (newSchema.oneOf) allowedKeys = allowedKeysForOneOf;

    for (let [key, value] of Object.entries(merged)) {
      let path = jsonPointerHelpers.append('', key);

      if (!allowedKeys.includes(key) || isExtension(key)) {
        yield {
          op: 'remove',
          path,
        };
      } else if (!currentKeys.has(key)) {
        yield {
          op: 'add',
          path,
          value,
        };
      } else if (value !== currentSchema[key]) {
        yield {
          op: 'replace',
          path,
          value,
        };
      }
    }
  }

  static applyShapePatch(
    schema: SchemaObject | null,
    patch: ShapePatch
  ): SchemaObject {
    const result = JsonPatch.applyPatch(
      schema,
      JsonPatch.deepClone([...OperationPatch.operations(patch)]),
      undefined,
      false // don't mutate the original schema
    );

    return result.newDocument!;
  }

  static isPolymorphic(schema: SchemaObject) {
    return !!(schema.allOf || schema.anyOf || schema.oneOf);
  }
}

function initialSchema(
  rootInput: any,
  openAPIVersion: SupportedOpenAPIVersions
): OpenAPIV3.SchemaObject {
  if (rootInput === null) {
    if (openAPIVersion === '3.0.x') return { nullable: true };
    // @ts-ignore we need to retype this as a union of 3 & 3.1
    return { type: 'null' };
  } else if (Array.isArray(rootInput)) {
    return {
      type: 'array',
      items: rootInput.length
        ? initialSchema(rootInput[0], openAPIVersion)
        : {},
    };
  } else if (typeof rootInput === 'object') {
    return { type: 'object' };
  } else if (typeof rootInput === 'string') {
    return { type: 'string' };
  } else if (typeof rootInput === 'number') {
    return { type: 'number' };
  } else if (typeof rootInput === 'boolean') {
    return { type: 'boolean' };
  } else {
    throw new Error('Could not learn JSON Schema');
  }
}

export const allowedMetaDataForAll: string[] = [
  'title',
  'nullable',
  'description',
  'example',
  'examples',
  'default',
  'deprecated',
  'externalDocs',
];

export const allowedKeysForOneOf: string[] = [
  ...allowedMetaDataForAll,
  'oneOf',
  'allOf',
  'not',
  'discriminator',
];

export const allowedKeysForObject: string[] = [
  ...allowedMetaDataForAll,
  'additionalProperties',
  'type',
  'maxProperties',
  'minProperties',
  'required',
  'properties',
];
export const allowedKeysForArray: string[] = [
  ...allowedMetaDataForAll,
  'type',
  'items',
  'maxItems',
  'minItems',
  'uniqueItems',
];

export const allowedKeysForString: string[] = [
  ...allowedMetaDataForAll,
  'type',
  'format',
  'pattern',
  'maxLength',
  'minLength',
];

export const allowedKeysForInteger: string[] = [
  ...allowedMetaDataForAll,
  'type',
  'minimum',
  'maximum',
  'multipleOf',
];

export const isExtension = (key: string) => key.startsWith('x-');
