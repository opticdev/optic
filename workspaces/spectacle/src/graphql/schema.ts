export const schema = `
scalar JSON
schema {
  query: Query
  mutation: Mutation
}
type Mutation {
  applyCommands(commands: [JSON], batchCommitId: ID, commitMessage: String, clientId: ID, clientSessionId: ID): AppliedCommandsResult
  startDiff(diffId: ID, captureId: ID): StartDiffResult
  invalidateCaches: InvalidateCachesResult
}
type InvalidateCachesResult {
  batchCommitId: ID
}
type AppliedCommandsResult {
  batchCommitId: ID
}
type StartDiffResult {
  notificationsUrl: String
  listDiffsQuery: String
  listUnrecognizedUrlsQuery: String
}

"""
Queries for Spectacle
"""
type Query {  
  # All HTTP requests defined in the spec
  requests: [HttpRequest]

  # URI paths for endpoints
  paths: [Path]
  
  # Shape choices based on shape ID
  shapeChoices(shapeId: ID): [OpticShape]
  
  # Endpoint changes since a given batch commit ID
  endpointChanges(sinceBatchCommitId: String): EndpointChanges
  
  # All batch commits defined in a spec
  batchCommits: [BatchCommit]
  
  # Diffs for existing endpoints and unrecognized URLs
  diff(diffId: ID): DiffState

  endpoint(pathId: ID, method: String): Endpoint

  # Metadata about the current spec
  metadata: SpecMetadata
}

"""
SpecMetadata - misc data about the spec
"""
type SpecMetadata {
  id: String
}

type DiffState {
  diffs: JSON
  unrecognizedUrls: JSON
}
type Path {
  # URL path pattern without parameter names
  absolutePathPattern: String
  
  # URI path pattern with parameter names included 
  absolutePathPatternWithParameterNames: String
  
  # Whether or not the path component is a parameter or not
  isParameterized: Boolean
  
  # Parent path ID
  parentPathId: String
  
  # Name of path or name of parameter if parameterized
  name: String
  
  # Path ID
  pathId: String
}

type Endpoint {
  commands: EndpointCommands
}

type EndpointCommands {
  remove: [JSON]
}

"""
HTTP Body, which could be for a request or response
"""
type HttpBody {
  # Content type of the given HTTP body
  contentType: String
  
  # Root shape ID for the HTTP body. Look at the shapeChoices query getting more information about the root shape
  rootShapeId: String
}

"""
HTTP Request
"""
type HttpRequest {
  id: ID
  
  # Path components for the HTTP request
  pathComponents: [PathComponent]
  
  # URL path pattern without parameter names
  absolutePathPattern: String
  
  # URI path pattern with parameter names included 
  absolutePathPatternWithParameterNames: String
  
  # Path ID
  pathId: ID
  
  # HTTP method for the HTTP request
  method: String
  
  # Request bodies associated with this HTTP request
  bodies: [HttpBody]
  
  # Responses associated with this HTTP request
  responses: [HttpResponse]
  
  # Changes for the HTTP request based on the give batch commit ID
  changes(sinceBatchCommitId: String): ChangesResult
  
  # Path contributions which define descriptions
  pathContributions: JSON
  
  # Request contributions which define descriptions
  requestContributions: JSON
}

"""
Path Component
The URL /users/{id} would have three path components, one for the root, one for users and one for the parameter id.
"""
type PathComponent {
  id: ID
  
  # Name of the path component or path parameter
  name: String
  
  # Defines whether or not the path component is parameterized
  isParameterized: Boolean
  
  # Path component contributions which define descriptions
  contributions: JSON
}

"""
HTTP Response
"""
type HttpResponse {
  id: ID
  
  # HTTP status code for the HTTP response
  statusCode: Int
  
  # Response bodies associated with this HTTP response
  bodies: [HttpBody]
  
  # Changes for the HTTP response based on the give batch commit ID
  changes(sinceBatchCommitId: String): ChangesResult
  
  # HTTP response contributions which define descriptions
  contributions: JSON
}

"""
Object Field Metadata, which provides information about a field
"""
type ObjectFieldMetadata {
  name: String
  
  # Field ID for the given field
  fieldId: ID
  
  # Changes for the field based on the give batch commit ID
  changes(sinceBatchCommitId: String): ChangesResult
  
  # Shape ID of field. Use the shapeChoice query to get shape information.
  shapeId: ID
  
  # Field contributions which define descriptions
  contributions: JSON
}

"""
Object Metadata, which provides information about an object
"""
type ObjectMetadata {
  # Fields for the given object
  fields: [ObjectFieldMetadata]
}

"""
Array Metadata, which provides information about an array
"""
type ArrayMetadata {
  # Changes for the array based on the give batch commit ID
  changes(sinceBatchCommitId: String): ChangesResult
  
  # Shape ID of the array. Use the shapeChoice query to get shape information.
  shapeId: ID
}

"""
Optic Shape
"""
type OpticShape {
  id: ID
  
  # JSON type of shape. Could be string, number, boolean, null, object, or array.
  jsonType: String
  
  # Metadata if jsonType is object
  asObject: ObjectMetadata
  
  # Metadata if jsonType is array
  asArray: ArrayMetadata
  
  # changes(sinceBatchCommitId: String): ChangesResult
  # exampleValue: [JSON]
}

"""
Change Result, which defines whether if something was added or updated
"""
type ChangesResult {
  # Whether or not the change was one that was added
  added: Boolean
  
  # Whether or not the change was one that was updated
  changed: Boolean
}

"""
Endpoint Changes
"""
type EndpointChanges {
  # URL for Optic change documentation
  opticUrl: String
  
  # Changed endpoints for the batch commit ID provided to query
  endpoints: [EndpointChange]
}

"""
Endpoint Change, which provides information about an endpoint change found since the given batch commit ID
"""
type EndpointChange {
  # Metadata about the endpoint change
  change: EndpointChangeMetadata
  
  # Path ID related to the given change
  pathId: String
  
  # Absolute path pattern with parameters
  path: String
  
  # HTTP method for the given endpoint change
  method: String
  
  # Descriptions for the endpoint change
  contributions: JSON
}

"""
Endpoint Change Metadata
"""
type EndpointChangeMetadata {
  # Defines how the endpoint changed
  category: String
}

"""
Batch Commit, which provides information about changes in the spec captured as batch commits
"""
type BatchCommit {
  # When the batch commit was created
  createdAt: String
  
  # Batch commit ID
  batchId: String
  
  # Batch commit message
  commitMessage: String
}
`;
