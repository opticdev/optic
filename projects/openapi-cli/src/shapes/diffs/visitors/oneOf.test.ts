import { diffValueBySchema } from '..';
import { objectOrStringOneOf } from '../../../tests/fixtures/oneof-schemas';
import { SchemaObject } from '../../schema';

describe('one of json schema diff visitor', () => {
  const jsonSchema: SchemaObject = {
    type: 'object',
    properties: {
      polyProp: {
        oneOf: [{ type: 'string' }, { type: 'number' }],
      },
    },
  };

  it('when valid, no diff', () => {
    const input = {
      polyProp: 'hello-string',
    };
    const diffs = diffValueBySchema(input, jsonSchema);
    expect([...diffs]).toHaveLength(0);
  });
  it('when valid case 2, no diff', () => {
    const input = {
      polyProp: 123,
    };
    const diffs = diffValueBySchema(input, jsonSchema);
    expect([...diffs]).toHaveLength(0);
  });

  it('when new primitive types provided to existing one of ', () => {
    const input = {
      polyProp: true,
    };

    const diffs = diffValueBySchema(input, jsonSchema);

    // expect(result.totalDiffsAfterPatches).toBe(0);
    expect([...diffs]).toMatchSnapshot();
  });

  it('when new field in one of object variant of one of', () => {
    const jsonSchema: SchemaObject = {
      type: 'object',
      properties: {
        polyProp: {
          oneOf: [{ type: 'object', properties: {} }, { type: 'number' }],
        },
      },
    };

    const input = {
      polyProp: { hello: 'world' },
    };

    const diffs = diffValueBySchema(input, jsonSchema);

    expect([...diffs]).toMatchSnapshot();
  });

  it('when root schema is obejct and is shown an array', () => {
    const jsonSchema: SchemaObject = {
      type: 'object',
      properties: {
        sup: { type: 'string' },
      },
    };

    const input: any = [];

    const diffs = diffValueBySchema(input, jsonSchema);
    expect([...diffs]).toMatchSnapshot();
  });

  it('can add an additional branch to a complex one of', () => {
    const jsonSchema: SchemaObject = objectOrStringOneOf();

    const input: any = {
      location: {
        principality: {
          city: 'San Fransisco',
          population: 830000,
          coordinates: [1, 2, 3],
        },
      },
    };

    const diffs = diffValueBySchema(input, jsonSchema);
    expect([...diffs]).toMatchSnapshot();
  });
});
