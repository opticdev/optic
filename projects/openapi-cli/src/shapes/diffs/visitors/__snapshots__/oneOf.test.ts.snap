// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one of json schema diff visitor can add an additional branch to a complex one of 1`] = `
Array [
  Object {
    "example": Array [
      1,
      2,
      3,
    ],
    "instancePath": "/location/principality/coordinates",
    "key": "0",
    "keyword": "type",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/location/properties/principality/properties/coordinates/oneOf/0",
  },
  Object {
    "example": Array [
      1,
      2,
      3,
    ],
    "instancePath": "/location/principality/coordinates",
    "key": "1",
    "keyword": "type",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/location/properties/principality/properties/coordinates/oneOf/1",
  },
  Object {
    "example": Array [
      1,
      2,
      3,
    ],
    "instancePath": "/location/principality/coordinates",
    "key": "oneOf",
    "keyword": "oneOf",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/location/properties/principality/properties/coordinates/oneOf",
  },
]
`;

exports[`one of json schema diff visitor when new field in one of object variant of one of 1`] = `
Array [
  Object {
    "example": Object {
      "polyProp": Object {
        "hello": "world",
      },
    },
    "instancePath": "/polyProp/hello",
    "key": "hello",
    "keyword": "additionalProperties",
    "kind": "AdditionalProperty",
    "parentObjectPath": "/properties/polyProp/oneOf/0/additionalProperties",
    "propertyExamplePath": "/polyProp/hello",
    "propertyPath": "/properties/polyProp/oneOf/0/additionalProperties/hello",
  },
  Object {
    "example": Object {
      "hello": "world",
    },
    "instancePath": "/polyProp",
    "key": "1",
    "keyword": "type",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/polyProp/oneOf/1",
  },
  Object {
    "example": Object {
      "hello": "world",
    },
    "instancePath": "/polyProp",
    "key": "oneOf",
    "keyword": "oneOf",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/polyProp/oneOf",
  },
]
`;

exports[`one of json schema diff visitor when new primitive types provided to existing one of  1`] = `
Array [
  Object {
    "example": true,
    "instancePath": "/polyProp",
    "key": "0",
    "keyword": "type",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/polyProp/oneOf/0",
  },
  Object {
    "example": true,
    "instancePath": "/polyProp",
    "key": "1",
    "keyword": "type",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/polyProp/oneOf/1",
  },
  Object {
    "example": true,
    "instancePath": "/polyProp",
    "key": "oneOf",
    "keyword": "oneOf",
    "kind": "UnmatchedType",
    "propertyPath": "/properties/polyProp/oneOf",
  },
]
`;

exports[`one of json schema diff visitor when root schema is obejct and is shown an array 1`] = `
Array [
  Object {
    "example": Array [],
    "instancePath": "",
    "key": "",
    "keyword": "type",
    "kind": "UnmatchedType",
    "propertyPath": "",
  },
]
`;
